File: java/Week1/src/DigitsTest.java
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.stream.Stream;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;

public class DigitsTest {

    private static Stream<Arguments> AddDigitsParameterProvider() {
        return Stream.of(
          Arguments.of(29, 11),
          Arguments.of(123, 6),
          Arguments.of(527, 14),
          Arguments.of(123456, 21)
        );
    }

    @ParameterizedTest
    @MethodSource("AddDigitsParameterProvider")
    public void TestAddDigits(int n, int expected)
    {
        int actual = Digits.AddDigits(n);
        assertEquals(expected, actual);
    }

    private static Stream<Arguments> LargestNumberParameterProvider() {
        return Stream.of(
          Arguments.of(2, 99),
          Arguments.of(3, 999),
          Arguments.of(4, 9999),
          Arguments.of(5, 99999)
        );
    }

    @ParameterizedTest
    @MethodSource("LargestNumberParameterProvider")
    public void TestLargestNumber(int n, int expected)
    {
        int actual = Digits.LargestNumber(n);
        assertEquals(expected, actual);
    }


    private static Stream<Arguments> OptimalEqualSplitParameterProvider() {
        return Stream.of(
          Arguments.of(3, 10, 9),
          Arguments.of(1, 2, 2),
          Arguments.of(10, 5, 0),
          Arguments.of(4, 15, 12)
        );
    }

    @ParameterizedTest
    @MethodSource("OptimalEqualSplitParameterProvider")
    public void TestOptimalEqualSplit(int n, int m, int expected)
    {
        int actual = Digits.OptimalEqualSplit(n, m);
        assertEquals(expected, actual);
    }

    private static Stream<Arguments> PersistenceParameterProvider() {
        return Stream.of(
          Arguments.of(39, 3),
          Arguments.of(4, 0),
          Arguments.of(25, 2),
          Arguments.of(999, 4)
        );
    }

    @ParameterizedTest
    @MethodSource("PersistenceParameterProvider")
    public void TestPersistence(int n, int expected)
    {
        int actual = Digits.Persistence(n);
        assertEquals(expected, actual);
    }
    
}

File: csharp/Week1.Test/DigitsTest.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;

namespace Week1.Test
{
    [TestClass]
    public class DigitsTest
    {
        [DataTestMethod]
        [DataRow(29,11)]
        [DataRow(123,6)]
        [DataRow(527,14)]
        [DataRow(123456,21)]
        public void TestAddDigits(int n, int expectedSum)
        {
            Assert.AreEqual(expectedSum, Digits.AddDigits(n));
        }

        [DataTestMethod]
        [DataRow(2, 99)]
        [DataRow(3, 999)]
        [DataRow(4, 9999)]
        [DataRow(5, 99999)]
        public void TestLargestNumber(int n, int expectedResult)
        {
            Assert.AreEqual(expectedResult, Digits.LargestNumber(n));
        }

        [DataTestMethod]
        [DataRow(3, 10, 9)]
        [DataRow(1, 2, 2)]
        [DataRow(10, 5, 0)]
        [DataRow(4, 4, 4)]
        [DataRow(4, 15, 12)]
        public void TestOptimalEqualSplit(int n, int m, int expectedResult)
        {
            Assert.AreEqual(expectedResult, Digits.OptimalEqualSplit(n,m));
        }

        [DataTestMethod]
        [DataRow(39, 3)]
        [DataRow(4, 0)]
        [DataRow(25,2)]
        [DataRow(999, 4)]
        public void TestPersistence(int n, int expectedResult)
        {
            Assert.AreEqual(expectedResult, Digits.Persistence(n));
        }


    }
}
